test/LaunchDarkly.EventSource.Tests/Stubs/StubHandler.cs

```diff
@@ -26,41 +28,154 @@ public class StubMessageHandler : HttpMessageHandler
-            _responses.Enqueue(response);
+    internal class StubResponseWithHttpResponse : StubResponse
+    {
+        private readonly HttpResponseMessage message;
+
+        public StubResponseWithHttpResponse(HttpResponseMessage message)
+        {
+            this.message = message;
+        }
+
+        override public HttpResponseMessage MakeResponse(CancellationToken cancellationToken)
+        {
+            return message;
+        }
+    }
 
-        public void QueueStringResponse(string content)
-        {
-            var response =
-                new HttpResponseMessage(HttpStatusCode.OK)
-                {
-                    Content = new StringContent(content, System.Text.Encoding.UTF8, "text/event-stream")
-                };
+    internal class StubResponseWithIOError : StubResponse
+    {
+        override public HttpResponseMessage MakeResponse(CancellationToken cancellationToken)
+        {
+            throw new HttpRequestException("Unit Test Exception Message");
+        }
+    }
 
-            _responses.Enqueue(response);
+    internal class StubResponseWithStream : StubResponse
+    {
+        private readonly StreamAction[] _actions;
+
+        public StubResponseWithStream(StreamAction[] actions)
+        {
+            _actions = actions;
+        }
+
+        override public HttpResponseMessage MakeResponse(CancellationToken cancellationToken)
+        {
+            var httpResponse = new HttpResponseMessage(HttpStatusCode.OK);
+            var streamRead = new AnonymousPipeServerStream(PipeDirection.In);
+            var streamWrite = new AnonymousPipeClientStream(PipeDirection.Out, streamRead.ClientSafePipeHandle);
+            var content = new StreamContent(streamRead);
+            content.Headers.ContentType = new MediaTypeHeaderValue("text/event-stream");
+            httpResponse.Content = content;
+
+            Task.Run(() => WriteStreamingResponse(streamWrite, cancellationToken));
+
+            return httpResponse;
+        }
+        
+        private async Task WriteStreamingResponse(Stream output, CancellationToken cancellationToken)
+        {
+            try
+            {
+                foreach (var action in _actions)
+                {
+                    if (action.Delay != TimeSpan.Zero)
+                    {
+                        await Task.Delay(action.Delay, cancellationToken);
+                    }
+                    if (action.ShouldQuit())
+                    {
+                        return;
+                    }
+                    byte[] data = Encoding.UTF8.GetBytes(action.Content);
+                    await output.WriteAsync(data, 0, data.Length, cancellationToken);
+                }
+                // if we've run out of actions, leave the stream open until it's cancelled
+                while (true)
+                {
+                    await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken);

```

Comment:
Something here that I'm unsure about: my expectation was that if the client cancels the request, `cancellationToken` will be cancelled, causing this line to throw an exception so that we'll exit the loop. However, in my debugging I was never able to see it get to line 143. I'm not sure if that means that I'm wrong about how the cancellation token works for HTTP requests, or if it just doesn't work because this is not a real HTTP server, or if it is somehow shutting down this task as soon as the unit test finishes so we never get that far.

So it is possible that these tasks are not stopping. I would rather not accumulate a bunch of old background tasks during a test run, just on general principle. On the other hand, they do go away as soon as the test run is done, so I'm not sure it is a huge problem.